### PR TITLE
Added option to directly use a custom key id for disk encryption set.

### PR DESCRIPTION
**Pull request by shanoor**
## PR Checklist

- [ ] I have added example(s) inside the [./examples/] folder
- [ ] I have added the example(s) to the integration test list for [normal (~30 minutes)](./workflows/standalone-scenarios.json) or [long runner >30 minutes](./workflows/standalone-scenarios-longrunners.json)
- [x] I have checked the [coding conventions as per the wiki](https://github.com/aztfmod/terraform-azurerm-caf/wiki)
- [x] I have checked to ensure there aren't other open Pull Requests for the same update/change?

## Description

Added option to directly use a custom key id for disk encryption set.

## Does this introduce a breaking change

- [ ] YES
- [x] NO

## Testing

```
disk_encryption_sets = {
  set1 = {
    name               = "DSE001"
    resource_group_key = "dse"
    key_vault_key_id   = "https://keyvault001.vault.azure.net/keys/dse-cmk/fec2d7f7b8014cf0980551ce30dfec57"
  }
}
```

### CODE REVIEW

File: disk_encryption_sets.tf [hunk 1]
```diff
@@ -8,8 +8,8 @@ module "disk_encryption_sets" {
...
   base_tags           = try(local.global_settings.inherit_tags, false) ? try(local.combined_objects_resource_groups[try(each.value.resource_group.lz_key, local.client_config.landingzone_key)][try(each.value.resource_group.key, each.value.resource_group_key)].tags, {}) : {}
-  key_vault_key_id    = local.combined_objects_keyvault_keys[try(each.value.keyvault_key.lz_key, local.client_config.landingzone_key)][try(each.value.key_vault_key_key, each.value.key_vault_key.key)].id
-  keyvault_id         = local.combined_objects_keyvaults[try(each.value.keyvault.lz_key, local.client_config.landingzone_key)][each.value.keyvault.key].id
+  key_vault_key_id    = try(each.value.key_vault_key_id, local.combined_objects_keyvault_keys[try(each.value.keyvault_key.lz_key, local.client_config.landingzone_key)][try(each.value.key_vault_key_key, each.value.key_vault_key.key)].id)
```
Comment (arnaudlh) [inline disk_encryption_sets.tf:11]:
```suggestion
  key_vault_key_id    = can(each.value.key_vault_key_id) ?  each.value.key_vault_key_id : local.combined_objects_keyvault_keys[try(each.value.keyvault_key.lz_key, local.client_config.landingzone_key)][try(each.value.key_vault_key_key, each.value.key_vault_key.key)].id)
```

**Review by arnaudlh — CHANGES_REQUESTED** [review]
Please check if the refactoring for ```key_vault_key_id``` is working for you. 

This is to force the evaluation of ```key_vault_key_id``` at plan time (hence more stable and no issue when you add/delete keys).

File: disk_encryption_sets.tf (hunk 1 above)
Comment (shanoor) [reply disk_encryption_sets.tf]: I made the change, it works just fine 👍

File: disk_encryption_sets.tf [hunk 2]
```diff
@@ -8,8 +8,8 @@ module "disk_encryption_sets" {
...
   base_tags           = try(local.global_settings.inherit_tags, false) ? try(local.combined_objects_resource_groups[try(each.value.resource_group.lz_key, local.client_config.landingzone_key)][try(each.value.resource_group.key, each.value.resource_group_key)].tags, {}) : {}
-  key_vault_key_id    = local.combined_objects_keyvault_keys[try(each.value.keyvault_key.lz_key, local.client_config.landingzone_key)][try(each.value.key_vault_key_key, each.value.key_vault_key.key)].id
-  keyvault_id         = local.combined_objects_keyvaults[try(each.value.keyvault.lz_key, local.client_config.landingzone_key)][each.value.keyvault.key].id
+  key_vault_key_id    = can(each.value.key_vault_key_id) ? each.value.key_vault_key_id : local.combined_objects_keyvault_keys[try(each.value.keyvault_key.lz_key, local.client_config.landingzone_key)][try(each.value.key_vault_key_key, each.value.key_vault_key.key)].id
+  keyvault_id         = try(each.value.key_vault_key_id, null) == null ? local.combined_objects_keyvaults[try(each.value.keyvault.lz_key, local.client_config.landingzone_key)][each.value.keyvault.key].id : null
```
Comment (LaurentLesle) [inline disk_encryption_sets.tf:12]: We are trying to leverage the can() when adding an option to reference the id of an object. Let me know if the following is working for you?

```suggestion
  keyvault_id         = can(each.value.key_vault_key_id) ? null : local.combined_objects_keyvaults[try(each.value.keyvault.lz_key, local.client_config.landingzone_key)][each.value.keyvault.key].id
```

File: modules/security/disk_encryption_set/keyvault_policy.tf [hunk 3]
```diff
@@ -1,4 +1,5 @@
 resource "azurerm_key_vault_access_policy" "des" {
+  count        = try(var.keyvault_id, null) == null ? 0 : 1
```
Comment (LaurentLesle) [inline modules/security/disk_encryption_set/keyvault_policy.tf:2]: Suggest you update the variable keyvault_id like

```
variable keyvault_id {
 default = null
}
```

With that variable default set to null you don't need to add the try()
```suggestion
  count        = var.keyvault_id == null ? 0 : 1
```

**Review by LaurentLesle — CHANGES_REQUESTED** [review]
Thanks @shanoor for the PR. I have added some comments. Can you address them first?

File: modules/security/disk_encryption_set/keyvault_policy.tf (hunk 3 above)
Comment (shanoor) [reply modules/security/disk_encryption_set/keyvault_policy.tf]: I made the change, it works :)

**Review by arnaudlh — APPROVED** [review]
LGTM

File: modules/security/disk_encryption_set/keyvault_policy.tf
Comment (LaurentLesle) [inline modules/security/disk_encryption_set/keyvault_policy.tf:1]: I can see you are setting the keyvault_id optional. Are you planning to grant the disk encryption permission outside of this code? To get the key, the DES need to have a policy set to get, warp and unwarp the key

File: disk_encryption_sets.tf [hunk 4]
```diff
@@ -8,8 +8,8 @@ module "disk_encryption_sets" {
...
   base_tags           = try(local.global_settings.inherit_tags, false) ? try(local.combined_objects_resource_groups[try(each.value.resource_group.lz_key, local.client_config.landingzone_key)][try(each.value.resource_group.key, each.value.resource_group_key)].tags, {}) : {}
-  key_vault_key_id    = local.combined_objects_keyvault_keys[try(each.value.keyvault_key.lz_key, local.client_config.landingzone_key)][try(each.value.key_vault_key_key, each.value.key_vault_key.key)].id
-  keyvault_id         = local.combined_objects_keyvaults[try(each.value.keyvault.lz_key, local.client_config.landingzone_key)][each.value.keyvault.key].id
+  key_vault_key_id    = can(each.value.key_vault_key_id) ? each.value.key_vault_key_id : local.combined_objects_keyvault_keys[try(each.value.keyvault_key.lz_key, local.client_config.landingzone_key)][try(each.value.key_vault_key_key, each.value.key_vault_key.key)].id
```
Comment (LaurentLesle) [inline disk_encryption_sets.tf:11]: Suggest you add as well the nested variable
```suggestion
  key_vault_key_id    = can(each.value.key_vault_key_id) || can(each.value.key_vault_key.id) ? try(each.value.key_vault_key_id, each.value.key_vault_key.id) : local.combined_objects_keyvault_keys[try(each.value.keyvault_key.lz_key, local.client_config.landingzone_key)][try(each.value.key_vault_key_key, each.value.key_vault_key.key)].id
```

**Review by LaurentLesle — CHANGES_REQUESTED** [review]
See my comments

File: modules/security/disk_encryption_set/keyvault_policy.tf
Comment (shanoor) [reply modules/security/disk_encryption_set/keyvault_policy.tf:1]: Yes, we are in a situation where the key is stored in a keyvault managed by another team, so they are responsible for creating the policy, we just need to consume the key by using its direct URL.

File: disk_encryption_sets.tf (hunk 4 above)
Comment (shanoor) [reply disk_encryption_sets.tf]: Works for me

**Pull request merged**